### PR TITLE
Failure alerts  Seldon Core and Prometheus integration

### PR DESCRIPTION
**Pull request by i-chvets**
For more details refer to https://github.com/canonical/seldon-core-operator/issues/52

### CODE REVIEW

**Review by DnPlas — COMMENTED** [review]
Thanks @i-chvets, just some comments, looks good overall.

File: tests/integration/test_charm.py [hunk 1]
```diff
@@ -84,23 +87,125 @@ async def test_seldon_istio_relation(ops_test: OpsTest):
...
+
+    # obtain alert rules from Prometheus
+    rules_url = f"http://{prometheus_url}:9090/api/v1/rules"
+    alert_rules_result = await fetch_url(rules_url)
+
+    # verify that all Seldon alert rules are in the list
+    assert alert_rules_result is not None
+    assert alert_rules_result["status"] == "success"
+    rules = alert_rules_result["data"]["groups"][0]["rules"]
+    assert rules[0]["name"] == "SeldonWorkqueueTooManyRetries"
```
[the remark this answers is not in the document]
Comment (i-chvets) [reply tests/integration/test_charm.py:149]: Hmm, it should not. I will re-do it in safer manner.

File: tests/integration/test_charm.py [hunk 2]
```diff
@@ -84,23 +87,125 @@ async def test_seldon_istio_relation(ops_test: OpsTest):
...
+        sdep["metadata"]["name"] = "seldon-model-1"
+        client.create(sdep, namespace=namespace)
+    assert_available(client, seldon_deployment, "seldon-model-1", namespace)
+
+    # remove deployment that was created by Seldon, reconcile alert will fire
+    client.delete(Deployment, name="seldon-model-1-example-0-classifier", namespace=namespace)
+
+    # Prometheus scraping is done once a minute
+    # wait for 2 minutes to ensure alert state is propagated
+    for min_idx in range(2):
```
[the remark this answers is not in the document]
Comment (i-chvets) [reply tests/integration/test_charm.py:184]: Tenacity did not work with `async` URL fetching. I will review it one more time.

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -84,23 +87,125 @@ async def test_seldon_istio_relation(ops_test: OpsTest):
...
+
+async def test_seldon_alert_rules(ops_test: OpsTest):
+    """Test Seldon alert rules."""
+    # NOTE: This test is re-using deployments created in test_build_and_deploy()
+    namespace = ops_test.model.name
+    client = Client()
+
+    # setup Prometheus
+    prometheus = "prometheus-k8s"
+    await ops_test.model.deploy(prometheus, channel="latest/edge", trust=True)
```
[the remark this answers is not in the document]
Comment (i-chvets) [reply tests/integration/test_charm.py:122]: Don't we want to pull latest all the time like we do for other components?

File: tests/integration/test_charm.py [hunk 4]
```diff
@@ -84,23 +87,125 @@ async def test_seldon_istio_relation(ops_test: OpsTest):
     stop=tenacity.stop_after_attempt(30),
     reraise=True,
 )
-def assert_available(client, seldon_deployment, namespace):
+def assert_available(client, seldon_deployment, deploy_name, namespace):
```
[the remark this answers is not in the document]
Comment (i-chvets) [reply tests/integration/test_charm.py:90]: Will do `seldon_deployment_name`.

File: tests/integration/test_charm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (i-chvets) [reply tests/integration/test_charm.py:149]: Done.

File: tests/integration/test_charm.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (i-chvets) [reply tests/integration/test_charm.py:184]: Made it work with `tenacity`.

File: tests/integration/test_charm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (i-chvets) [reply tests/integration/test_charm.py:142]: Will switch to `stable`

File: tests/integration/test_charm.py [hunk 5]
```diff
@@ -84,23 +86,158 @@ async def test_seldon_istio_relation(ops_test: OpsTest):
...
+        verbs=None,
+    )
+    with open("examples/serve-simple-v1.yaml") as f:
+        sdep = seldon_deployment(yaml.safe_load(f.read()))
+        sdep["metadata"]["name"] = "seldon-model-1"
+        client.create(sdep, namespace=namespace)
+    assert_available(client, seldon_deployment, "seldon-model-1", namespace)
+
+    # remove deployment that was created by Seldon, reconcile alert will fire
+    client.delete(Deployment, name="seldon-model-1-example-0-classifier", namespace=namespace)
```
Comment (DnPlas) [inline tests/integration/test_charm.py:219]: I think this delete call must refer to the `SeldonDeployment` instead, you can use the resource class from the previous lines `seldon_deployment`

Comment (i-chvets) [reply tests/integration/test_charm.py:219]: Not quite. This is exactly the trick I am playing on Seldon to trigger the alert. SeldonDeployment still exists, but classifier deployment `seldon-model-1-example-0-classifier` is removed. This causes reconciliation, which in turn triggers reconcile error alert.
If I remove SeldonDeployment, no alerts will be triggered.

Comment (DnPlas) [reply tests/integration/test_charm.py:219]: Makes sense.

**Review by DnPlas — APPROVED** [review]
LGTM, thanks @i-chvets

**Pull request merged**